File: homing-studio-workspace/src/main/java/hue/captains/singapura/js/homing/studio/workspace/NavigatorSecretaryModule.java
package hue.captains.singapura.js.homing.studio.workspace;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The Secretary for the Studio Workspace's {@code NavigatorParty} — the
 * navigation message bus. Per the Diligent Secretaries doctrine, a pure
 * {@code (state, envelope) → Step} routing function plus an initial state,
 * exported as a single {@code NavigatorSecretary} JS object.
 *
 * <p><b>Just redirects.</b> The Secretary holds no UI and renders nothing.
 * Its sole job is to relay a tree selection to every Party member: an
 * incoming {@code NodeSelected} is rebroadcast as {@code NavigateTo}. The
 * {@link TreeWidget} is the source actor; any widget that joins the Party
 * (a future content / detail pane) receives the redirect and decides what
 * to do with it. Decoupling the producer from the consumer is the whole
 * point — the tree never names who consumes its selections.</p>
 *
 * <h2>State shape</h2>
 * <pre>{@code
 * {
 *     lastSelected : node | null,        // most recent selection (diligence)
 *     recentUnknown: [{ kind, from }]    // last 10 unrecognised messages, bounded
 * }
 * }</pre>
 *
 * <h2>Message kinds</h2>
 * <ul>
 *   <li>{@code NodeSelected(node)} — record + {@code BroadcastToMembers}
 *       a {@code NavigateTo(node)}.</li>
 *   <li>Anything else — append to {@code recentUnknown} (bounded at 10);
 *       no action.</li>
 * </ul>
 *
 * @since homing-studio-workspace — studio navigation Party
 */
public record NavigatorSecretaryModule() implements DomModule<NavigatorSecretaryModule> {

    /** The single export — a JS object with {@code initial} and {@code behavior}. */
    public record NavigatorSecretary() implements Exportable._Constant<NavigatorSecretaryModule> {}

    public static final NavigatorSecretaryModule INSTANCE = new NavigatorSecretaryModule();

    @Override
    public ImportsFor<NavigatorSecretaryModule> imports() {
        return ImportsFor.<NavigatorSecretaryModule>builder().build();
    }

    @Override
    public ExportsOf<NavigatorSecretaryModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new NavigatorSecretary()));
    }
}

File: homing-studio-workspace/src/main/resources/homing/js/hue/captains/singapura/js/homing/studio/workspace/NavigatorSecretaryModule.js
// =============================================================================
// NavigatorSecretaryModule — the Secretary for the Studio Workspace's
// NavigatorParty (the navigation message bus).
//
// Per the Diligent Secretaries doctrine: a pure (state, envelope) -> Step
// behavior plus an initial state, exposed as members of one top-level object
// for clean framework import + isolated test-harness load. Side-effect-free;
// no DOM, no console, no captures.
//
// "Just redirects": an incoming NodeSelected (from the tree) is rebroadcast
// to every Party member as NavigateTo; an incoming NodeOpened (Enter /
// double-click — the intentional open) is rebroadcast as OpenDoc. Two tiers,
// same redirect shape: select is cheap + frequent (panes that re-label), open
// is expensive + deliberate (the content pane that fetches + renders). The
// Secretary names no consumer — any widget that joins receives the redirect.
// Diligence: it remembers the last selection AND the last open, and bounds a
// recentUnknown ring for observability via inspect().
// =============================================================================

var NavigatorSecretary = {

    initial: {
        lastSelected:  null,
        lastOpened:    null,
        recentUnknown: []
    },

    /**
     * Pure routing function. (state, envelope) -> { newState, actions }.
     * Same inputs always produce same outputs; no side effects.
     */
    behavior: function (state, envelope) {
        var msg = envelope.message;

        switch (msg.kind) {

            case "NodeSelected": {
                // Redirect: relay the tree's selection to all members. Cheap,
                // high-frequency — fires on every arrow-move / single click.
                return {
                    newState: {
                        lastSelected:  msg.node,
                        lastOpened:    state.lastOpened,
                        recentUnknown: state.recentUnknown
                    },
                    actions: [{
                        kind:    "BroadcastToMembers",
                        message: { kind: "NavigateTo", node: msg.node }
                    }]
                };
            }

            case "NodeOpened": {
                // Redirect: relay an INTENTIONAL open (Enter / double-click) to
                // all members as OpenDoc. The content pane reacts to this — and
                // only this — so the expensive doc render is paid on intent.
                return {
                    newState: {
                        lastSelected:  state.lastSelected,
                        lastOpened:    msg.node,
                        recentUnknown: state.recentUnknown
                    },
                    actions: [{
                        kind:    "BroadcastToMembers",
                        message: { kind: "OpenDoc", node: msg.node }
                    }]
                };
            }

            default: {
                // Diligent defensive observability: track unknowns (bounded at
                // 10 via slice(-10)) so they surface via inspect() rather than
                // vanishing silently.
                var recent = state.recentUnknown.concat([{
                    kind: msg.kind,
                    from: envelope.from
                }]).slice(-10);
                return {
                    newState: {
                        lastSelected:  state.lastSelected,
                        lastOpened:    state.lastOpened,
                        recentUnknown: recent
                    },
                    actions: []
                };
            }
        }
    }
};
